### DISCUSSION TITLE
Clickhouse-keeper digest authentication "username:password". How to?

### DISCUSSION THREAD

**Discussion by Stanislav-Vlasov** [ClickHouse/ClickHouse#55783] [Q&A] [answered]
The oficial Keeper docs says about support for digest authentication scheme uses the pair username:password. https://clickhouse.com/docs/en/guides/sre/keeper/clickhouse-keeper. I am looking for examples how to configure that "username:password" authentication scheme on Keeper server side. I didn't find any relevant documentation for that purpose.

I am considering two usage scenarios.
 
The first scenario, the client in order to access the keeper server performs authentication with a login password. Anonymous access must be prohibited, the keeper must not accept the connection without provided login and password by client.
The second scenario, all Keeper nodes inside the same Keeper cluster must authenticatе earch other befor internal cluster replication to avoid any anonymous access.

I know that the Кeeper can protect connections with a certificate. But this requires an unnecessarily complex approach and infrastructure planning. And not necessary for Test and Stage environments. So, login and password should be o'key for that purpose and easily configurable.

**Comment by Stanislav-Vlasov** [comment]
Maybe the developers will answer whether it is implemented or not? I think if the documentation describes that feature, but in fact it is not implemented, perhaps it would be more correct for me to write a proposal?

**Comment by antonio2368** [comment, marked as answer]
Hi, sorry for not answering. I didn't notice the question earlier.

Auth is supported in Keeper, it should be fully compatible with ZooKeeper's authentication described here: https://zookeeper.apache.org/doc/r3.5.2-alpha/zookeeperProgrammers.html#sc_ZooKeeperAccessControl

In the documentation, we add mostly Keeper-specific things while most ZooKeeper features should work with Keeper.
We relied on many different already existing sources like ZooKeeper or ZooKeeper client documentation which we should definitely improve and minimise the need to investigate how a feature should be used.

I also recommend to look at our integration tests to see how auth is used with Kazoo client: https://github.com/ClickHouse/ClickHouse/blob/master/tests/integration/test_keeper_auth/test.py#L65

**Comment by Catwoolfii** [comment]
Hello!

I understand that the "getAcl" and "addauth" commands are not supported for client cli?
I tried with zookeeper cli_mt, but it doesn't work...

**Reply by antonio2368** [reply]
Could you give me an example of a case which doesn't work for you?

**Reply by Catwoolfii** [reply]
I have a three-node clickhouse-keeper test installation:
```
root@ch-04:~# sh -c 'echo mntr | curl telnet://localhost:9181' 2>&1 | grep zk_version
zk_version      v24.1.3.31-stable-135b08cbd28a5832e9e70c3b7d09dd4134845ed3
```
next, I connect to the "cli_mt" client and try to execute these commands:
```
root@ch-04:~# /usr/lib/zookeeper/bin/cli_mt -h 127.0.0.1:9181
Watcher SESSION_EVENT state = CONNECTED_STATE
Got a new session id: 0x35

ls /
time = 0 msec
/: rc = 0
        keeper
        clickhouse
time = 0 msec

getAcl /clickhouse

addauth digest user:1234qwer
```
Only first command was executed. Maybe I missed something?

**Reply by antonio2368** [reply]
I don't think those commands are supported by `cli_mt` (https://github.com/apache/zookeeper/blob/master/zookeeper-client/zookeeper-client-c/README#L76) and sadly not by our `keeper-client`.
Try with python client maybe like in our integration tests?

**Reply by Catwoolfii** [reply]
Yes, it seems that it is:
```
root@ch-04:~# /usr/lib/zookeeper/bin/cli_mt -h 127.0.0.1:9181
Watcher SESSION_EVENT state = CONNECTED_STATE
Got a new session id: 0x38

help
    create [+[e|s|c|t=ttl]] <path>
    create2 [+[e|s|c|t=ttl]] <path>
    delete <path>
    set <path> <data>
    get <path>
    ls <path>
    ls2 <path>
    sync <path>
    exists <path>
    wexists <path>
    myid
    verbose
    addauth <id> <scheme>
    config
    reconfig [-file <path> | -members <serverId=host:port1:port2;port3>,... |  -add <serverId=host:port1:port2;port3>,... | -remove <serverId>,...] [-version <version>]
    quit
```
I was inattentive...
Since I am interested in the cli, the question is settled. Thanks for help.

**Reply by ilkarataev** [reply]
So I will be answer to my self.
CLI client clickhouse-keeper can't use auth data, and you can't check it.

in clickhouse-keeper you can add         <identity>keeper:example_pass</identity>
example_pass = string(in docs said that we need base64 I test and it's wrong)

```
<yandex>
    <zookeeper>
        <node>
            <host>chk-nodes-0.local.</host>
            <port>2181</port>
        </node>
        <node>
            <host>chk-nodes-1.local.</host>
            <port>2181</port>
        </node>
        <node>
            <host>chk-nodes-2.local.</host>
            <port>2181</port>
        </node>
        <identity>keeper:Th0OoiSb</identity>
    </zookeeper>
    <distributed_ddl>
        <path>/clickhouse/clickhouse88/task_queue/ddl</path>
    </distributed_ddl>
</yandex>```
```
In clickhouse server you need

```
<yandex>
    <zookeeper>
        <node>
            <host>chk-nodes-0.local.</host>
            <port>2181</port>
        </node>
        <node>
            <host>chk-nodes-1.local.</host>
            <port>2181</port>
        </node>
        <node>
            <host>chk-nodes-2.local.</host>
            <port>2181</port>
        </node>
        <identity>keeper:example_pass</identity>
    </zookeeper>
    <distributed_ddl>
        <path>/clickhouse/clickhouse88/task_queue/ddl</path>
    </distributed_ddl>
</yandex>
```

But client cli connected without auth)Very fun) I don't know why)

karvet@karvet-Latitude-7420:~$ clickhouse-keeper-client -h example server -p2181
Connected to ZooKeeper at 95.217.171.145:2181 with session_id 57
Keeper feature flag FILTERED_LIST: enabled
Keeper feature flag MULTI_READ: enabled
Keeper feature flag CHECK_NOT_EXISTS: disabled
Keeper feature flag CREATE_IF_NOT_EXISTS: disabled
/ :)

**Reply by antonio2368** [reply]
try getting data for a node that was created by CH using the identity.
You should get:
```
/ :) get /clickhouse
Coordination error: Not authenticated, path /clickhouse
```

**Reply by ilkarataev** [reply]
How I can auth with identity?

**Comment by Dnivd** [comment]
Has anything been clarified since the last message?

**Reply by ilkarataev** [reply]
use cli from zookeper he can auth.

**Comment by Dnivd** [comment]
**Thanks for the idea, I'll think about a use case.**

But it is still more desirable to use it within the framework of clickhouse-keeper, otherwise it turns out to be a hodgepodge of everything.
It's strange that it hasn't been implemented so far, it's like an increase in security.

p.s. And it's not very clear how to specify and in which config the variables about what will be used will be diger

**Comment by Dnivd** [comment]
Apparently, there is no desire to maintain information security in clickhouse, unfortunately, this task is one of the last for developers to implement(
